/*
 * Processes related to genome and annotations
 */

process BUILD_STAR {
  conda 'envs/star.yml'
  label 'maxCPU'

  input:
    tuple path(genomeFasta), path(genomeGtf)
  output:
    path("STAR_REF/*")
  script:
    """
STAR --runThreadN ${task.cpus} \
--runMode genomeGenerate \
--limitGenomeGenerateRAM 35000000000 \
--genomeSAindexNbases 14 \
--genomeDir STAR_REF \
--genomeFastaFiles ${ genomeFasta } \
--sjdbGTFfile ${ genomeGtf } \
--sjdbOverhang 149
    """

}

/*
 * Aligned reads to both the genome and transciptome generated by STAR
 */


process ALIGN_STAR {
  conda 'envs/star.yml'
  label 'maxCPU'

  input:
    tuple path(STAR_INDEX_DIR), val(sampleID), path(forward_read), path(reverse_read)
  output:
    tuple val(sampleID), \
          path("${ forward_read.simpleName }Aligned.sortedByCoord.out.bam"), \
          path("${ reverse_read.simpleName }Aligned.sortedByCoord.out.bam"), emit: genomeMapping \
    tuple val(sampleID), \
          path("${ forward_read.simpleName }Aligned.toTranscriptome.out.bam"), \
          path("${ reverse_read.simpleName }Aligned.toTranscriptome.out.bam"), emit:transcriptomeMapping
  script:
    """
    STAR --twopassMode Basic \
    --limitBAMsortRAM 65000000000 \
    --outFilterType BySJout \
    --outSAMunmapped Within \
    --genomeDir ${ STAR_INDEX_DIR } \
    --outSAMattributes NH HI AS NM MD MC \
    --outFilterMismatchNoverReadLmax 0.04 \
    --outFilterMismatchNmax 999 \
    --outFilterMultimapNmax 20 \
    --alignIntronMin 20 \
    --alignSJoverhangMin 8 \
    --alignMatesGapMax 1000000 \
    --alignIntronMax 1000000 \
    --alignSJDBoverhangMin 1 \
    --sjdbScore 1 \
    --outSAMtype BAM SortedByCoordinate \
    --runThreadN ${ task.cpus } \
    --readFilesCommand zcat \
    --quantMode TranscriptomeSAM \
    --outFileNamePrefix ${ sampleID } \
    --readFilesIn ${ forward_read } ${ reverse_read }
    """

}
